File: backend/app/engines/expiry.py
"""Drop supplier quotes that have already expired.

Indexed reuse and homepage deals must not outlive Duffel's expires_at.
If the field is missing, the observation TTL still applies.
"""

from __future__ import annotations

from datetime import datetime, timezone

from app.models import Offer


def _parse_expiry(raw: str | None) -> datetime | None:
    if not raw:
        return None
    text = raw.strip()
    if not text:
        return None
    try:
        exp = datetime.fromisoformat(text.replace("Z", "+00:00"))
    except ValueError:
        return None
    if exp.tzinfo is None:
        exp = exp.replace(tzinfo=timezone.utc)
    return exp


def offer_unexpired(offer: Offer, now: datetime | None = None) -> bool:
    exp = _parse_expiry(offer.expires_at)
    if exp is None:
        return True
    moment = now or datetime.now(timezone.utc)
    if moment.tzinfo is None:
        moment = moment.replace(tzinfo=timezone.utc)
    return exp > moment


def unexpired(offers: list[Offer], now: datetime | None = None) -> list[Offer]:
    moment = now or datetime.now(timezone.utc)
    return [o for o in offers if offer_unexpired(o, moment)]

File: backend/src/engines/expiry.rs
use chrono::{DateTime, Utc};

use crate::models::Offer;

pub fn parse_offer_time(raw: Option<&str>) -> Option<DateTime<Utc>> {
    let text = raw?.trim();
    if text.is_empty() {
        return None;
    }
    let text = text.replace('Z', "+00:00");
    DateTime::parse_from_rfc3339(&text)
        .or_else(|_| DateTime::parse_from_rfc3339(&format!("{text}+00:00")))
        .ok()
        .map(|d| d.with_timezone(&Utc))
        .or_else(|| {
            chrono::NaiveDateTime::parse_from_str(&text, "%Y-%m-%dT%H:%M:%S")
                .ok()
                .map(|n| n.and_utc())
        })
}

pub fn offer_unexpired(offer: &Offer) -> bool {
    offer_unexpired_at(offer, None)
}

pub fn offer_unexpired_at(offer: &Offer, now: Option<DateTime<Utc>>) -> bool {
    let Some(exp) = parse_offer_time(offer.expires_at.as_deref()) else {
        return true;
    };
    let moment = now.unwrap_or_else(Utc::now);
    exp > moment
}

pub fn unexpired(offers: &[Offer]) -> Vec<Offer> {
    unexpired_at(offers, None)
}

pub fn take_unexpired(offers: Vec<Offer>) -> Vec<Offer> {
    offers.into_iter().filter(offer_unexpired).collect()
}

pub fn unexpired_at(offers: &[Offer], now: Option<DateTime<Utc>>) -> Vec<Offer> {
    let moment = now.unwrap_or_else(Utc::now);
    offers
        .iter()
        .filter(|o| offer_unexpired_at(o, Some(moment)))
        .cloned()
        .collect()
}
